052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,96 @@
+It would also need confirming that all resource creates/updates go through `serverSideApply`, not something that should be done in this proposal.
+
+And Kubernetes recommend it is used on CREATE also:
+[From](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#L272):
+
+```
+ReconcileResult<T> result = ReconcileResult.created(operation().inNamespace(namespace).resource(desired).create());
+```
+
+To:
+
+```
+ReconcileResult<T> result = ReconcileResult.patched(operation().inNamespace(namespace).withName(name).patch(new PatchContext.Builder().withPatchType(PatchType.SERVER_SIDE_APPLY).withForce(true).build(), desired);
+```
+
+According to the [package being used](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+NOTE: The version of fabric8io being used is [6.6.1](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/pom.xml#L106) which [contains Server Side Apply](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+The operators will need modifying in terms of change discovery so that they only check fields that are owned by themselves, `metadata.managedFields`.
+
+The [`ResourceDiff`](https://github.com/strimzi/strimzi-kafka-operator/blob/c3522cf4b17004004a676854d37ba01bb9a44800/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceDiff.java#L46-L78) will need to filter only the parts of the resource that Strimzi owns as to not trigger a reconcile loop for externally owned resource paths.
+This will allow the removal of `ignorableFields` throughout the codebase as paths are being selected now, rather than filtering out fields.
+
+It appears [this is a place](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ServiceOperator.java#L123) that the `desired` state imports existing external annotations, this would need to change to not include existing external annotations within the `desired` state as then Kubernetes would create an ownership conflict with Server Side Apply.
+
+
+### Modifications to existing behaviours
+
+The only change to existing (non-Strimzi specific) behaviours is that Strimzi won't care about other fields it chooses not to care about. This does mean that another entity within the cluster can modify things without Strimzi's knowledge, but it shouldn't matter if the `metadata.managedFields` are chosen correctly.
+
+I do not know the codebase for the strimzi kafka operator so cannot comment on specific behaviours that might change, as this is just a proposal I suspect that is better placed in the eventual PR if this proposal gets accepted.
```

Comment:
I think in this section we need more clarity about how this interacts with things like the template mechanism. Because I think this feature provides a different way to achieve the same thing. In other words I can use pod templates and secret templates etc to configure some of aspects of the derived resources. Or I can not use the template mechanism and instead let the operator create those derived resources and then tinker as I like with whatever fields the operator doesn't own. We need to give users a clear idea of which is the right mechanism for them to use and in what circumstances. 

Just to be clear: This isn't about the codebase (i.e. the Java code), it's about the CR API, which is the main contract between the user and the operator. Obviously there's a balance to be struck, but if we don't define here the scope of this work there's the risk that the PR gets bogged down/never merged after a lot of effort (both yours and ours) has been expended on implementation. No one wants that.

Reply:
>  We need to give users a clear idea of which is the right mechanism for them to use and in what circumstances.

Fair. I think my starting suggestion would be:
- If you want to configure something for a single object (e.g. a single Kafka cluster), and it's in scope of the configuration, use that
- If you want to configure something that's not the currently configurable fields, you can use other mechanisms to do so (Kyverno, other controllers), and Strimzi won't interfere
- If "you" must configure something through some other method, and it's not directly being set in a different way by Strimzi, then Strimzi won't interfere and it'll "just work" (e.g. cattle annotations, pvc resizing annotations, metallb annotations. The user didn't make a choice about this other than the "I want to run Strimzi on this cluster that uses those technologies))
- If "you" (and again this may not be the direct user of Strimzi, but the platform/corporate/enterprise team enforcing policy on a wide scale) want to set certain fields, then again if it doesn't conflict, go for it. (e.g., The [Datadog admission controller](https://docs.datadoghq.com/containers/cluster_agent/admission_controller/?tab=operator) setting labels, Kyverno setting labels or annotations or automated topologySpreadConstraints)

I guess it's worth setting out explicitly if you have two things that both want a completely incompatible change (idk, Strimzi want's `replicas: 3` and you have something else saying `replicas: 1`), you're going to have a Bad Time™️.
(And that's not a change, you'll currently have a bad time)

Reply:
TBH, I think that:
* Users should always prefer to configure things in the Strimzi custom resources as that is where they are created. It should be always preferred to validate the resources if you want some uniformity or enforce some features then to inject it in some admission controller. 
* Things which are not configurable in Strimzi are often not configurable for a reason. So if it is not configurable, think 10 times if you really should work around it.